### PR TITLE
feat: add zkSync with RedStone tutorial

### PR DESCRIPTION
**Pull request by cehali**
# Tutorial submission

## Title of the tutorial
zkSync and RedStone stable price NFT marketplace tutorial
- 

## Brief tutorial description

This tutorial will show how to use RedStone oracles on the zkSync Era blockchain. It is designed as an example dApp of a stable price NFT marketplace. It means that sellers can create sell orders (offers), specifying the price amount in USD. But buyers are able to pay with native coins, the required amount of which is calculated dynamically at the moment of the order execution. Users will be requested to add a few crucial parts which will make it possible to use RedStone data on-chain and run application on zkSync local testnet.

- 

## Checklist
Please ensure you've completed the following tasks:

- [x] I confirm that this tutorial is an original work and hasn't been published elsewhere.
- [x] I confirm that I am the rightful intellectual property owner (author) of this submission.
- [x] The tutorial includes referenced and relevant external sources (if applicable).
- [x] The tutorial includes working code snippets to illustrate concepts.
- [x] I have adhered to the tutorial guidelines provided in the repository.

### CODE REVIEW

File: tutorials/zkSync-RedStone-stable-price-marketplace-tutorial/README.md [hunk 1]
```diff
@@ -0,0 +1,402 @@
...
+
+We've used hardhat test framework to contract tests. All the tests are located in the [test](test/) folder.
+
+## 🔥 Tutorial how to integrate RedStone oracles on zkSync
+
+### Prepare repo
+
+#### 1. Clone this repo
+```sh
+git clone https://github.com/redstone-finance/stable-price-marketplace
```
Comment (idea404) [inline tutorials/zkSync-RedStone-stable-price-marketplace-tutorial/README.md:100]:
```suggestion
git clone https://github.com/zkSync-Community-Hub/tutorials
```

File: tutorials/zkSync-RedStone-stable-price-marketplace-tutorial/README.md [hunk 2]
```diff
@@ -0,0 +1,402 @@
...
+cd stable-price-marketplace
```
Comment (idea404) [inline tutorials/zkSync-RedStone-stable-price-marketplace-tutorial/README.md:101]:
```suggestion
cd tutorials/tutorials/zkSync-RedStone-stable-price-marketplace-tutorial
```

File: tutorials/zkSync-RedStone-stable-price-marketplace-tutorial/README.md [hunk 3]
```diff
@@ -0,0 +1,402 @@
...
+
+You can check the code of the React app in the `src` folder. We tried to simplify it as much as possible and leave only the core marketplace functions.
+
+The main UI logic is located in the `App.tsx` file, and the contract interaction logic is in the `blockchain.ts` file.
+
+If you take a look into the `blockchain.ts` file code, you'll notice that each contract call that needs to process RedStone data is made on a contract instance, that was wrapped by [@redstone-finance/evm-connector](https://www.npmjs.com/package/@redstone-finance/evm-connector).
+
+### Tests
+
+We've used hardhat test framework to contract tests. All the tests are located in the [test](test/) folder.
```
Comment (idea404) [inline tutorials/zkSync-RedStone-stable-price-marketplace-tutorial/README.md:92]: Do the tests all pass on your end? When I run the tests I get:

```bash
yarn test
yarn run v1.22.19
$ NODE_ENV=test hardhat test --network zkSyncTestnet


  ExampleNFT
    ✔ Should deploy ExampleNFT (2136ms)
    ✔ Should mint 2 NFTs (1778ms)
    ✔ User1 should approve spending and User2 should spend (2838ms)

  Marketplace core functions test
    ✔ Should deploy contracts (3038ms)
    ✔ Should mint NFT (889ms)
    ✔ Seller should post sell order for token 1 with ETH price (2859ms)
    ✔ Should get all orders
Expected ETH amount: 1.0
    ✔ Buying should fail with smaller amount then seller requested (316ms)
Expected ETH amount: 1.0
    ✔ Buyer should buy token 1 for ETH price (917ms)
    ✔ Should post and cancel order (4811ms)

  Marketplace core functions test
    ✔ Should deploy contracts (2953ms)
    ✔ Should mint NFT (1901ms)
    ✔ Seller should post sell order for token 2 with stable USD price (2778ms)
    ✔ Should wrap marketplace contract with redstone wrapper
Expected ETH amount: 0.0
    1) Buying should fail with smaller amount then seller requested
Expected ETH amount: 0.0
    2) Buyer should buy token for USD price expressed in ETH


  14 passing (31s)
  2 failing

  1) Marketplace core functions test
       Buying should fail with smaller amount then seller requested:
     AssertionError: Expected transaction to be reverted
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (node:internal/process/task_queues:96:5)
      at async Context.<anonymous> (test/stable-marketplace.test.ts:88:5)

  2) Marketplace core functions test
       Buyer should buy token for USD price expressed in ETH:
     Error: cannot estimate gas; transaction may fail or may require manual gas limit [ See: https://links.ethers.org/v5-errors-UNPREDICTABLE_GAS_LIMIT ] (error={"reason":"execution reverted","code":"UNPREDICTABLE_GAS_LIMIT","method":"estimateGas","transaction":{"from":"0xa61464658AfeAf65CccaaFD3a512b69A83B77618","gasPrice":{"type":"BigNumber","hex":"0x0ee6b280"},"to":"0x0a67078A35745947A37A552174aFe724D8180c25","value":{"type":"BigNumber","hex":"0x00"},"data":"0xd96a094a0000000000000000000000000000000000000000000000000000000000000000455448000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000025171e97dd018a845a3920000000200000010184df94c1cdf3dacc79d3081d78a5798364ee89ab96a76a9145e796ad9db325299c10dba57d6b4cb294ea811069d44477a21c718d6785174d52fa86bd6ef94a1b00013136393434333736373338343623302e322e352372656473746f6e652d6d61696e2d64656d6f000026000002ed57011e0000","type":0,"accessList":null},"error":{"reason":"processing response error","code":"SERVER_ERROR","body":"{\"jsonrpc\":\"2.0\",\"error\":{\"code\":3,\"message\":\"execution reverted\",\"data\":\"0x\"},\"id\":68}","error":{"code":3,"data":"0x"},"requestBody":"{\"method\":\"eth_estimateGas\",\"params\":[{\"gasPrice\":\"0xee6b280\",\"type\":\"0x0\",\"value\":\"0x0\",\"from\":\"0xa61464658afeaf65cccaafd3a512b69a83b77618\",\"to\":\"0x0a67078a35745947a37a552174afe724d8180c25\",\"data\":\"0xd96a094a0000000000000000000000000000000000000000000000000000000000000000455448000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000025171e97dd018a845a3920000000200000010184df94c1cdf3dacc79d3081d78a5798364ee89ab96a76a9145e796ad9db325299c10dba57d6b4cb294ea811069d44477a21c718d6785174d52fa86bd6ef94a1b00013136393434333736373338343623302e322e352372656473746f6e652d6d61696e2d64656d6f000026000002ed57011e0000\"}],\"id\":68,\"jsonrpc\":\"2.0\"}","requestMethod":"POST","url":"http://localhost:3050"}}, tx={"data":"0xd96a094a0000000000000000000000000000000000000000000000000000000000000000455448000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000025171e97dd018a845a3920000000200000010184df94c1cdf3dacc79d3081d78a5798364ee89ab96a76a9145e796ad9db325299c10dba57d6b4cb294ea811069d44477a21c718d6785174d52fa86bd6ef94a1b00013136393434333736373338343623302e322e352372656473746f6e652d6d61696e2d64656d6f000026000002ed57011e0000","to":{},"from":"0xa61464658AfeAf65CccaaFD3a512b69A83B77618","value":{"type":"BigNumber","hex":"0x00"},"type":0,"gasPrice":{},"nonce":{},"gasLimit":{},"chainId":{}}, code=UNPREDICTABLE_GAS_LIMIT, version=abstract-signer/5.7.0)
      at Logger.makeError (node_modules/@ethersproject/logger/src.ts/index.ts:269:28)
      at Logger.throwError (node_modules/@ethersproject/logger/src.ts/index.ts:281:20)
      at /Users/dennis/Documents/Code/tutorials/tutorials/zkSync-RedStone-stable-price-marketplace-tutorial/node_modules/@ethersproject/abstract-signer/src.ts/index.ts:301:31
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (node:internal/process/task_queues:96:5)
      at async Promise.all (index 7)



error Command failed with exit code 2.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

File: tutorials/zkSync-RedStone-stable-price-marketplace-tutorial/README.md [hunk 4]
```diff
@@ -0,0 +1,402 @@
...
+
+```js
+import { WrapperBuilder } from "@redstone-finance/evm-connector";
+
+async function getContractInstance(contractName) {
+  ...
+  return new ethers.Contract(address, abi, signer);
+}
+
+async function buy(orderId) {
```
Comment (idea404) [inline tutorials/zkSync-RedStone-stable-price-marketplace-tutorial/README.md:163]:
```suggestion
async function buy(orderId: string) {
```

File: tutorials/zkSync-RedStone-stable-price-marketplace-tutorial/README.md [hunk 5]
```diff
@@ -0,0 +1,402 @@
...
+### Test dApp locally
+
+#### 1. Check if tests pass
+```sh
+yarn test
+``` 
+
+#### 2. Compile contracts
+```sh
+yarn compile
```
Comment (idea404) [inline tutorials/zkSync-RedStone-stable-price-marketplace-tutorial/README.md:206]: This command does not exist in `package.json`

File: tutorials/zkSync-RedStone-stable-price-marketplace-tutorial/README.md [hunk 6]
```diff
@@ -0,0 +1,402 @@
...
+
+<img alt="my-nfts" src="docs/img/my-nfts.png" width="300" />
+
+##### Post sell orders
+Once you mint any NFTs, you can post sell order for each one of them. Click the SELL button and provide the USD value. You will be asked to confirm 2 transactions: for NFT transfer approval, and for the marketplace order creation. After their confirmation, you will see your order in the Orders column.
+
+<img alt="orders" src="docs/img/orders.png" width="300" />
+
+##### Buy NFTs
+You can also switch metamask account and buy the NFT. I would recommend to open the developer tools in browser at the network tab and explore network requests that are being sent before the buy transaction sending.
```
Comment (idea404) [inline tutorials/zkSync-RedStone-stable-price-marketplace-tutorial/README.md:258]: Sadly I try and get the following error:

image

Only one request is sent:

image

**Review by idea404 — CHANGES_REQUESTED** [review]
Some minor suggestions and comments

File: tutorials/zkSync-RedStone-stable-price-marketplace-tutorial/README.md (hunk 1 above)
Comment (cehali) [reply tutorials/zkSync-RedStone-stable-price-marketplace-tutorial/README.md]: Fixed

File: tutorials/zkSync-RedStone-stable-price-marketplace-tutorial/README.md (hunk 2 above)
Comment (cehali) [reply tutorials/zkSync-RedStone-stable-price-marketplace-tutorial/README.md]: Fixed

File: tutorials/zkSync-RedStone-stable-price-marketplace-tutorial/README.md (hunk 3 above)
Comment (cehali) [reply tutorials/zkSync-RedStone-stable-price-marketplace-tutorial/README.md]: Yes, but they will pass when you populate functions indicated in the tutorial like `buy` in dApp or `_getPriceFromOrder` in `StableMarketplace` contract

File: tutorials/zkSync-RedStone-stable-price-marketplace-tutorial/README.md (hunk 5 above)
Comment (cehali) [reply tutorials/zkSync-RedStone-stable-price-marketplace-tutorial/README.md]: Fixed

File: tutorials/zkSync-RedStone-stable-price-marketplace-tutorial/README.md (hunk 6 above)
Comment (cehali) [reply tutorials/zkSync-RedStone-stable-price-marketplace-tutorial/README.md]: Fixed, the problem was with an invalid function name in the tutorial text

**Review by idea404 — APPROVED** [review]

**Pull request merged**